### ISSUE TITLE
Pros and cons of using vmac_xmit_base

### ISSUE THREAD

**Issue by VasumathyJ** [acassen/keepalived#1957]
Hello

Using vmac_xmit_base seems to break the RFC3768 which states that the VRRP advertisements must be sent out with VRRP MAC as src address.
However I see some links which state that using vmac_xmit_base is the recommended option.

**Can you pls formally elaborate on the pros and cons of using vmac_xmit_base and clarify if/why that is the recommended option**

https://keepalived-devel.narkive.com/fmMiPB5j/vmac-vmac-xmit-base-and-igmp

**Comment by pqarmitage** [comment]
https://keepalived-devel.narkive.com/fmMiPB5j/vmac-vmac-xmit-base-and-igmp is very old as is keepalvied v1.2.13, and the descriptions of what happens is somewhat out of date compared to current versions of keepalived.

What were the issues relating to `use_vmac`:
1. If `use_vmac` is not used, then it does not comply with  RFC5798 (the current VRRP RFC), since adverts are not transmitted with the correct MAC address
2. If `use_vmac` is specified but `vmac_xmit_base` is not specified, then when keepalived starts, IGMP join group messages are sent using the VMAC MAC address, and this causes the problem identified in the link above. The IGMP join group messages are sent since keepalived needs to listen to the 224.0.0.18 multicast group.
3. If `vmac_xmit_base` is set, the IGMP join group messages are sent from the parent interface of the VMAC, and so there is not a problem with switches updating their MAC address caches. However, adverts are sent from the parent interface, and so you have the same problem as point 1 above.

The obvious solution would be to listen for adverts on the parent interface (and so IGMP join group messages would be sent with that interface's MAC address), but to send adverts on the VMAC interface. The problem is that messages sent with a MAC address of another interface on the local system (i.e. adverts sent from another master) are not received on the local system.

So it used to be a compromise regarding which of the above was not going to comply with the RFC.

As a consequence of the above, I decided to implement a solution to issue 2 above. keepalived will now configure nftables (or iptables if nftables support is not available) to move the IGMP join group messages from the VMAC interface to the parent interface, so that the IGMP messages are sent with the MAC address of the parent interface. Since keepalived has requested the join group on the VMAC interface, the kernel delivers messages (adverts) sent to 224.0.0.18 to the socket listening on the VMAC interface, but the IGMP messages are actually sent to the network with the MAC of the parent interface, which means that IGMP snooping network switches will forward the adverts to the local system.

Based on the above, if you specify `use_vmac` and your keepalived is built with (preferably) nftables support, or alternatively iptables support, then IGMP join group messages will be sent with the MAC address of the parent interface, adverts will be sent with the MAC address of the VMAC, and adverts will be able to be received (on the VMAC interface). Consequently there is no longer a problem with MAC caches getting updated incorrectly due to the IGMP messages being sent with the MAC address of the VMAC interface.

References in the message linked to above and messages it references, regarding needing to set arp_ignore, arp_announce, arp_filter, rp_filter are no longer relevant since keepalived sets them as necessary.

So to summarise, if you are not running a very old version of keepalived and you want to comply with the VRRP RFC, set `use_vmac` and do not set `vmac_xmit_base`. Also, ensure that your keepalived is build with nftables support (or iptables support if nftables cannot be supported).
